htdocs/class/smarty/xoops_plugins/function.xoStats.php

```diff
@@ -0,0 +1,129 @@
+ * File:     function.xoStats.php
+ * Type:     function
+ * Name:     xoStats
+ * Purpose:  XOOPS Members Statistics
+ * Author:   Lio MJ <liomj83@gmail.com>
+ * Examples:
+ *  <{xoStats}>
+ *  Latest Member : <a href="<{$xoops_url}>/userinfo.php?uid=<{$latestuid}>"><{$latestmemberuname}></a><br>
+ *  Total Post    : <{$totalpost}><br>
+ *  Total User    : <{$totaluser}><br>
+ *  Total Online  : <{$totalonline}><br>
+ *  Registered Today     : <{$totalregisteredtoday}><br>
+ *  Registered Yesterday : <{$totalregisteredyesterday}><br>
+ *
+ *  These are the names of all the values that will be assigned for Smarty
+ *      $latestmembername - name of newest member
+ *      $latestmemberuname - uname of newest member
+ *      $latestuid - uid of newest member
+ *      $totalonline - total members online
+ *      $totalpost - total posts by all members
+ *      $totalregisteredtoday - number of members registered today
+ *      $totalregisteredyesterday  - number of members registered yesterday
+ *      $totaluser - total number of members
+ *
+ */
+
+/**
+ * @param array   $params
+ * @param \Smarty $smarty Smarty instance
+ *
+ * @return void
+ */
+function smarty_function_xoStats($params, &$smarty)
+{
+    $cache = new Cache('system');
+    $stats = $cache->cacheRead('xostats', 'regen_xoStats', 30);
+
+    foreach ($stats as $k => $v) {
+        $smarty->assign($k, $v);
+    }
+}
+
+function regen_xoStats()
+{
+    global $xoopsUser, $xoopsModule;
+
+    $stats = array();
+
+    /** @var \XoopsMemberHandler $memberHandler */
+    $memberHandler = xoops_getHandler('member');
+
+    // Getting Total Online Users
+    /** @var \XoopsOnlineHandler $online_handler */
+    $online_handler = xoops_getHandler('online');
+    // set gc probabillity to 10% for now..
+    if (mt_rand(1, 100) < 11) {
+        $online_handler->gc(300);
+    }
+    if (is_object($xoopsUser)) {
+        $uid   = $xoopsUser->getVar('uid');
+        $uname = $xoopsUser->getVar('uname');
+    } else {
+        $uid   = 0;
+        $uname = '';
+    }
+
+    $requestIp = IPAddress::fromRequest()->asReadable();
+    $requestIp = (false === $requestIp) ? '0.0.0.0' : $requestIp;
+    if (is_object($xoopsModule)) {
+        $online_handler->write($uid, $uname, time(), $xoopsModule->getVar('mid'), $requestIp);
+    } else {
+        $online_handler->write($uid, $uname, time(), 0, $requestIp);
+    }
+    $onlines = $online_handler->getAll();
+    if (empty($onlines)) {
+        $stats['totalonline'] = 0;
+    } else {
+        $totalonline   = count($onlines);
+        $stats['totalonline'] = $totalonline;
+    }
+
+    //Getting Total Registered Users
+    $level_criteria = new \Criteria('level', 0, '>');
+    $criteria = new \CriteriaCompo($level_criteria);
+    $criteria24 = new \CriteriaCompo($level_criteria);
+    $criteria48 = new \CriteriaCompo($level_criteria);
+    $totaluser = $memberHandler->getUserCount($level_criteria);
```

Comment:
Wouldn't it be better to have plural here, i.e. "$totalusers" ? 

Also, if we are introducing new variables, shouldn't we follow the PSR naming suggestions, i.e. "$totalUsers" instead of "$totalusers"

Or is this a Smarty convention that we want to follow for historical reasons?

Reply:
Smarty only dictates some things like the `function smarty_function_*` name. The non-dictatated names should all be camelCase now.